lib/cartopy/geodesic/_geodesic.pyx

```diff
@@ -0,0 +1,240 @@
+                       points.
+                       The starting point(s) from which to travel.
+
+            * endpoints - An n (or 1) by 2 numpy.ndarray, list or tuple of 
+                          lon-lat points.
+                          The point(s) to travel to.            
+
+        Returns:
+            An n by 3 np.ndarray of distances, and the (forward) azimuths of the
+            start and end points. 
+
+        """        
+
+        cdef int n_points, i
+        cdef double[:, :] pts, epts
+
+        # Create numpy arrays from inputs, and ensure correct shape. Note: 
+        # reshape(-1) returns a 1D array from a 0 dimensional array as required 
+        # for broadcasting.        
+        pts = np.array(points, dtype=np.float64).reshape((-1, 2))
+        epts =  np.array(endpoints, dtype=np.float64).reshape((-1, 2))
+
+        n_points = max(pts.shape[0], epts.shape[0])
+
+        # Broadcast any length 1 arrays to the correct size.        
+        try:
+            tmp = np.zeros((n_points, 2))
+            tmp[:, 0] += pts[:, 0]
+            tmp[:, 1] += pts[:, 1]
+
+            pts = tmp
+
+            tmp = np.zeros((n_points, 2))
+            tmp[:, 0] += epts[:, 0]
+            tmp[:, 1] += epts[:, 1]
+
+            epts = tmp
+
+        except ValueError:
+            raise ValueError("Inputs must have common length n or length one.")
+
+        cdef double[:, :] results = np.empty((n_points, 3))
+
+        with nogil:
+            for i in prange(n_points):
+
+                geod_inverse(self.geod, pts[i, 1], pts[i, 0], epts[i, 1],
+                             epts[i, 0], &results[i,0], &results[i,1], 
+                             &results[i,2])
+
+        return results
+
+    def circle(self, double lon, double lat, double radius, int n_samples=180,
+               endpoint=False):
+        """
+        Find a geodesic circle of given radius at a given point.
+
+        Args:
+
+            * lon - Longitude coordinate of the centre.
+
+            * lat - Latitude coordinate of the centre.
+
+            * radius - The radius of the circle (metres).
+
+        Kwargs:
+
+            * n_samples - Integer number of sample points of circle.
+
+            * endpoint - Boolean for whether to repeat endpoint at the end of 
+                         returned array.           
+
+        Returns:
+            An n_samples by 2 np.ndarray of evenly spaced lon-lat points on the 
+            circle. 
+
+        """   
+
+        cdef int i
+
+        # Put the input arguments into c-typed values.        
+        cdef double[:,:] center = np.array([lon, lat]).reshape((1, 2))
+        cdef double[:] radius_m = np.asarray(radius).reshape(1)
+
+        azimuths = np.linspace(360., 0., n_samples, 
+                               endpoint=endpoint).astype(np.double)
+
+        return self.direct(center, azimuths, radius_m)[:, 0:2]
+
+    def __dealloc__(self):
```

Comment:
For my own sanity, I'd be tempted to move this to closer to the __init__ - having a malloc there is something which always rings alarm bells, and having this nearer to the malloc would ease those concerns.